### PR TITLE
Extensions: Add magiccards_popups extension

### PR DESCRIPTION
**Pull request by Traneptora**
This PR adds the magiccards_popups extension to New XKit.

magiccards_popups is an extension to New XKit. It creates popup images for links to [http://magiccards.info/](http://magiccards.info/).
Note that it only creates popups for links to direct cards, not searches. It adds popups on your dashboard and official tumblr pages, but not on user blogs.

### CODE REVIEW

File: Extensions/magiccards_popups.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            }
+            var old_href = $atag.prop("href"), theMatch, image_source, replaced;
+            if (old_href){
+                if (old_href.match(/^.*?\/\/magiccards.info\/.*?\/.*?\/.*?\.html$/)){
+                    theMatch = /^.*?\/\/magiccards.info\/(.*?)\/(.*?)\/(.*?)\.html$/.exec(old_href);
+                    image_source = "http://magiccards.info/scans/" + theMatch[2] + "/" + theMatch[1] + "/" + theMatch[3] + ".jpg";
+                    XKit.extensions.magiccards_popups.addPopup($atag, image_source);
+                } else if (old_href.match(/^http:\/\/t.umblr.com\/redirect\?z=http%3A%2F%2Fmagiccards\.info%2F.*?%2F.*?%2F.*?\.html(&.*)*$/)){
+                    replaced = /^http:\/\/t.umblr.com\/redirect\?z=(http%3A%2F%2Fmagiccards\.info%2F.*?%2F.*?%2F.*?\.html)(&.*)*$/.exec(old_href);
+                    theMatch = /^.*?\/\/magiccards.info\/(.*?)\/(.*?)\/(.*?)\.html$/.exec(decodeURIComponent(replaced[1]));
```
Comment (nightpool) [inline Extensions/magiccards_popups.js]: There's lots of code duplication here (match/exec with duplicated regexes instead of using the result from match) and the split flow is really hard to follow. Can I suggest something like the following?

```
var href = $atag.href;
href = replace_tumblr_redirects(href);

// new RegExp used here inline for clarity, this should be a property on the extension object
var match = href.match(new RegExp("//magiccards.info/([^/]+)/([^/]+)/([^.]+).html"))

image_source = "http://magiccards.info/scans/" + match[2] + "/" + match[1] + "/" + match[3] + ".jpg";
```

where replace_tumblr_redirects is something like 
`return decodeUriComponent(url.match("https?://t.umblr.com/redirect\?z=([^&]+)")[1])`

**Review by nightpool — CHANGES_REQUESTED** [review]
Definitely fix linting errors (mostly whitespace stuff, can be fixed with `eslint --fix`), and probably some of the below comments as well.

**Comment by nightpool** [comment]
My other general comment is that you're not exactly listening for posts here, semantically. Maybe you'd be better off using something like a MutationObserver instead? Then you can catch magiccards popups anywhere, not just posts on the dashboard.

File: Extensions/magiccards_popups.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        }, function(){
+            $(this).children("img").last().css("visibility", "hidden");
+        });
+        $atag.append($img);
+        $atag.on("mousemove.magiccards_popups", function(event){
+            var x = event.clientX, y = event.clientY;
+            if (y > window.innerHeight / 2){
+                y -= 485;
+            }
+            $(this).children("img").last().css({
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: `$(this)` ends up being `$atag`, not `$img`. That's why I use `$(this).children("img").last()`. As for why not use `$atag`? I've had issues in the past with JS doing this. Event handlers like this unambiguously use `this` to refer to the event dispatcher element (in this case, the atag).

File: Extensions/magiccards_popups.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ * magiccards_popups is an extension to New XKit. It creates popup images for links to http://magiccards.info/
+ * Note that it only creates popups for links to direct cards, not searches.
+ * It adds popups on your dashboard and official tumblr pages, but not on user blogs.
+ */
+
+XKit.extensions.magiccards_popups = new Object({
+
+    addPopup: function($atag, image_source){
+        var $img = $("<img>");
+        $img.prop("src", image_source);
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: $img is a jQuery object, not a DOM object. This won't work.

File: Extensions/magiccards_popups.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
+//* TITLE magiccards.info Popups **//
+//* VERSION 0.1.2 **//
+//* DESCRIPTION    Creates visual popups for links to cards on magiccards.info **//
+//* DEVELOPER thebombzen **//
+//* FRAME false **//
+//* BETA true **//
+
+/* globals XKit */
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: Alright, will remove.

File: Extensions/magiccards_popups.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: rip

File: Extensions/magiccards_popups.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+/*
+ * magiccards_popups is an extension to New XKit. It creates popup images for links to http://magiccards.info/
+ * Note that it only creates popups for links to direct cards, not searches.
+ * It adds popups on your dashboard and official tumblr pages, but not on user blogs.
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: Will do.

File: Extensions/magiccards_popups.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        $img.prop("src", image_source);
+        $img.css({
+            "padding": "0px",
+            "border":  "0px",
+            "width": "312px",
+            "height": "445px",
+            "position": "fixed",
+            "visibility": "hidden",
+            "z-index": "300"
+        });
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: I tried this, but it did not work. I can try a bit harder.

File: Extensions/magiccards_popups.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            }
+            $(this).children("img").last().css({
+                "top": (y + 20) + 'px',
+                "left": (x + 20) + 'px'
+            });
+        });
+    },
+
+    addPopupsIfNecessary: function(){
+        if (XKit.interface.is_tumblr_page() !== true){
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: Out of curiosity, why?

File: Extensions/magiccards_popups.js (hunk 2 above)
Comment (nightpool) [inline Extensions/magiccards_popups.js]: yes, that's why you should use `$img.css`.....

File: Extensions/magiccards_popups.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        $("a").each(function(){
+            var $atag = $(this);
+            if ($atag.hasClass("magiccards_popups-tooltip")){
+                return;
+            }
+            var old_href = $atag.prop("href"), theMatch, image_source, replaced;
+            if (old_href){
+                if (old_href.match(/^.*?\/\/magiccards.info\/.*?\/.*?\/.*?\.html$/)){
+                    theMatch = /^.*?\/\/magiccards.info\/(.*?)\/(.*?)\/(.*?)\.html$/.exec(old_href);
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: That got me too on the first write-through. I don't know why they do that, but they do that.

File: Extensions/magiccards_popups.js (hunk 2 above)
Comment (nightpool) [inline Extensions/magiccards_popups.js]: I don't know what you mean by "I've had issues in the past with JS doing this" doing what? using scope?

File: Extensions/magiccards_popups.js (hunk 1 above)
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: I made a note to do that when I cleaned up the code... and then I forgot. Oops.

File: Extensions/magiccards_popups.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+    removePopups: function(){
+        $(".magiccards_popups-tooltip").each(function() {
+             XKit.extensions.magiccards_popups.removePopup($(this));
+        });
+    },
+
+    running: false,
+
+    run: function() {
```
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: Out of curiosity, why?

File: Extensions/magiccards_popups.js (hunk 8 above)
Comment (nightpool) [inline Extensions/magiccards_popups.js]: Mostly for clarity.

File: Extensions/magiccards_popups.js (hunk 2 above)
Comment (thebombzen) [inline Extensions/magiccards_popups.js]: I've had issues with a local anonymous function being unable to reference a local variable from a function that declared it. I'll try your suggested method, and see if that runs into an issue.

**Comment by thebombzen** [comment]
Bump? Can I get a review of this again?

File: Extensions/magiccards_popups.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+		this.running = false;
+		XKit.extensions.magiccards_popups.removePopups();
+		XKit.tools.remove_css("magiccards_popups");
+		XKit.post_listener.remove("magiccards_popups");
+	},
+
+	magiccards_regex: new RegExp("^https?://magiccards.info/([^/]+)/([^/]+)/([^.]+).html"),
+
+	addPopupsIfNecessary: function() {
+		$("a").each(function() {
```
Comment (hobinjk) [inline Extensions/magiccards_popups.js]: Okay, I wrote a multi-line comment on how I thought this line would be far too slow to be run on any new post but it turns out it runs in less than 10ms on average so 👍

**Review by hobinjk — APPROVED** [review]
This is great work! Thank you for the contribution. I have one nitpick. If you don't feel like fixing it just let me know and I can merge this as-is

File: Extensions/magiccards_popups.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+	},
+
+	destroy: function() {
+		this.running = false;
+		XKit.extensions.magiccards_popups.removePopups();
+		XKit.tools.remove_css("magiccards_popups");
+		XKit.post_listener.remove("magiccards_popups");
+	},
+
+	magiccards_regex: new RegExp("^https?://magiccards.info/([^/]+)/([^/]+)/([^.]+).html"),
```
Comment (hobinjk) [inline Extensions/magiccards_popups.js]: Wait, sorry. I'm testing the popups locally and the anchoring of this regex to the start of the string causes issues with Tumblr's tracking links of the form `http://t.umblr.com/redirect?z=https%3A%2F%2Fmagiccards.info%2Fod%2Fen%2F20.html&t=blahblahblahmorestuff`

**Review by hobinjk — CHANGES_REQUESTED** [review]

File: Extensions/magiccards_popups.js (hunk 12 above)
Comment (thebombzen) [reply Extensions/magiccards_popups.js]: That's the purpose of lines 47-56.

Comment (hobinjk) [reply Extensions/magiccards_popups.js]: Oh that's strange. It didn't work the first time I tried it but it's working flawlessly now. Sorry for my confusion

**Review by hobinjk — APPROVED** [review]

**Comment by hobinjk** [comment]
@nightpool r+?

**Comment by nightpool** [comment]
I'd prefer if this wasn't post-listener based, but if that's not the direction we want to go then i'm fine as is

**Comment by thebombzen** [comment]
@nightpool do you have a suggested alternative?

**Review by nightpool — APPROVED** [review]

**Comment by nightpool** [comment]
@thebombzen I said a couple times during review and development that this makes more sense as an MutationObserver-based plugin.

(whoops, forgot to dismiss my review)

**Comment by thebombzen** [comment]
I changed it to use a MutationObserver to check for new child nodes on the DOM.

**Comment by Traneptora** [comment]
I'm closing this PR as I don't believe it works anymore. Any new attempt should go through the Scryfall API, anyway.
